### PR TITLE
refacto(ellipsoid): move ellipsoidSizes from Coordinates to Ellipsoid

### PR DESCRIPTION
**Pull request by zarov**
No breaking changes here, final export is still the same.

I also took the opportunity to quickly move Ellipsoid into an ES6 class.

### CODE REVIEW

File: src/Core/Math/Ellipsoid.js [hunk 1]
```diff
@@ -1,182 +1,171 @@
...
-
-
 import * as THREE from 'three';
+import proj4 from 'proj4';
 import Coordinates from 'Core/Geographic/Coordinates';
 
-function Ellipsoid(size) {
-    // Constructor
-
+export const ellipsoidSizes = {
```
Comment (gchoqueux) [inline src/Core/Math/Ellipsoid.js]: could you convert to Vector3?

File: src/Core/Math/Ellipsoid.js [hunk 2]
```diff
@@ -1,182 +1,171 @@
...
+        this.rayon_3 = size.z;
 
-    this._radiiSquared = new THREE.Vector3(size.x * size.x, size.y * size.y, size.z * size.z);
+        this.size = new THREE.Vector3(size.x, size.y, size.z);
 
-    this._oneOverRadiiSquared = new THREE.Vector3(size.x === 0.0 ? 0.0 : 1.0 / (size.x * size.x),
-        size.y === 0.0 ? 0.0 : 1.0 / (size.y * size.y),
-        size.z === 0.0 ? 0.0 : 1.0 / (size.z * size.z));
-}
+        this._radiiSquared = new THREE.Vector3(size.x * size.x, size.y * size.y, size.z * size.z);
```
Comment (gchoqueux) [inline src/Core/Math/Ellipsoid.js]: `this._radiiSquared = new THREE.Vector3().multiplyVectors(this.size, this.size);`

File: src/Core/Math/Ellipsoid.js [hunk 3]
```diff
@@ -1,182 +1,171 @@
...
 
-    this.rayon_1 = size.x;
-    this.rayon_2 = size.y;
-    this.rayon_3 = size.z;
+const normal = new THREE.Vector3();
 
-    this.size = new THREE.Vector3(size.x, size.y, size.z);
+class Ellipsoid {
+    constructor(size = ellipsoidSizes) {
+        this.rayon_1 = size.x;
```
Comment (gchoqueux) [inline src/Core/Math/Ellipsoid.js]: could add test if all values doesn't equal to zero.

File: src/Core/Math/Ellipsoid.js [hunk 4]
```diff
@@ -1,182 +1,171 @@
...
+        this._radiiSquared = new THREE.Vector3(size.x * size.x, size.y * size.y, size.z * size.z);
 
-Ellipsoid.prototype.geodeticSurfaceNormal = function geodeticSurfaceNormal(cartesian, target = new THREE.Vector3()) {
-    target.set(
-        cartesian.x() * this._oneOverRadiiSquared.x,
-        cartesian.y() * this._oneOverRadiiSquared.y,
-        cartesian.z() * this._oneOverRadiiSquared.z);
-    return target.normalize();
-};
+        this._oneOverRadiiSquared = new THREE.Vector3(size.x === 0.0 ? 0.0 : 1.0 / (size.x * size.x),
```
Comment (gchoqueux) [inline src/Core/Math/Ellipsoid.js]: `this._oneOverRadiiSquared = new THREE.Vector3(1,1,1).divide(this._radiiSquared);`

File: src/Core/Math/Ellipsoid.js [hunk 5]
```diff
@@ -1,182 +1,171 @@
...
+            size.y === 0.0 ? 0.0 : 1.0 / (size.y * size.y),
+            size.z === 0.0 ? 0.0 : 1.0 / (size.z * size.z));
+    }
 
-Ellipsoid.prototype.geodeticSurfaceNormalCartographic = function geodeticSurfaceNormalCartographic(coordCarto, target = new THREE.Vector3()) {
-    const longitude = THREE.Math.degToRad(coordCarto.longitude());
-    const latitude = THREE.Math.degToRad(coordCarto.latitude());
-    const cosLatitude = Math.cos(latitude);
+    geodeticSurfaceNormal(cartesian, target = new THREE.Vector3()) {
+        target.set(
```
Comment (gchoqueux) [inline src/Core/Math/Ellipsoid.js]:
```js
return cartesian.xyz(target).multiply(this._oneOverRadiiSquared).normalize();
```

File: src/Core/Math/Ellipsoid.js [hunk 6]
```diff
@@ -1,182 +1,171 @@
...
+     * @param {number} position.x
+     * @param {number} position.y
+     * @param {number} position.z
+     * @param {Coordinate} [target] coordinate to copy result
+     * @returns {Coordinate} an object describing the coordinates on the reference ellipsoid, angles are in degree
+     */
+    cartesianToCartographic(position, target = new Coordinates('EPSG:4326', 0, 0, 0)) {
+        // for details, see for example http://www.linz.govt.nz/data/geodetic-system/coordinate-conversion/geodetic-datum-conversions/equations-used-datum
+        // TODO the following is only valable for oblate ellipsoid of revolution. do we want to support triaxial ellipsoid?
+        const R = Math.sqrt(position.x * position.x + position.y * position.y + position.z * position.z);
```
Comment (gchoqueux) [inline src/Core/Math/Ellipsoid.js:73]:
```js
const R = position.length();
```

File: src/Core/Math/Ellipsoid.js [hunk 7]
```diff
@@ -1,182 +1,171 @@
...
+    intersection(ray) {
+        var EPSILON = 0.0001;
+        var O_C = ray.origin;
+        var dir = ray.direction;
+        // normalizeVector( dir );
 
-    var d = ((b * b) - (4 * a * c));
-    if (d < 0 || a === 0 || b === 0 || c === 0) { return false; }
+        var a =
+            ((dir.x * dir.x) / (this.size.x * this.size.x)) + ((dir.y * dir.y) / (this.size.y * this.size.y)) + ((dir.z * dir.z) / (this.size.z * this.size.z));
```
Comment (gchoqueux) [inline src/Core/Math/Ellipsoid.js]: could you use `this._radiiSquared`?

**Review by gchoqueux — CHANGES_REQUESTED** [review]
Good work.
I add some comments to simplify vector calculations but I don't verify the results.

File: src/Core/Math/Ellipsoid.js (hunk 6 above)
Comment (zarov) [reply src/Core/Math/Ellipsoid.js:73]: It doesn't go through unit test, or I need to modify `Coordinates`; as I am currently rewriting this class in another PR, I will adapt this later, not here.

**Comment by zarov** [comment]
I updated the code following some instructions of your review.

File: src/Core/Math/Ellipsoid.js (hunk 7 above)
Comment (zarov) [reply src/Core/Math/Ellipsoid.js]: done

File: src/Core/Math/Ellipsoid.js (hunk 5 above)
Comment (zarov) [reply src/Core/Math/Ellipsoid.js]: done

File: src/Core/Math/Ellipsoid.js (hunk 4 above)
Comment (zarov) [reply src/Core/Math/Ellipsoid.js]: I couldn't really change that: see my updated proposal.

File: src/Core/Math/Ellipsoid.js (hunk 3 above)
Comment (zarov) [reply src/Core/Math/Ellipsoid.js]: I don't see the test to add, sorry.

File: src/Core/Math/Ellipsoid.js (hunk 2 above)
Comment (zarov) [reply src/Core/Math/Ellipsoid.js]: done

File: src/Core/Math/Ellipsoid.js (hunk 1 above)
Comment (zarov) [reply src/Core/Math/Ellipsoid.js]: done

File: src/Core/Math/Ellipsoid.js [hunk 8]
```diff
@@ -1,182 +1,162 @@
...
+    intersection(ray) {
+        var EPSILON = 0.0001;
+        var O_C = ray.origin;
+        var dir = ray.direction;
+        // normalizeVector( dir );
 
-    var d = ((b * b) - (4 * a * c));
-    if (d < 0 || a === 0 || b === 0 || c === 0) { return false; }
+        var a =
+            ((dir.x * dir.x) / this._radiiSquared.x) + ((dir.y * dir.y) / this._radiiSquared.y) + ((dir.z * dir.z) / this._radiiSquared.z);
```
Comment (gchoqueux) [inline src/Core/Math/Ellipsoid.js]: sorry use this.`this._invRadiiSquared` instead of `this._radiiSquared`

Comment (zarov) [reply src/Core/Math/Ellipsoid.js]: Done, and rebased.

**Review by gchoqueux — APPROVED** [review]
